changeset/manager.go

```diff
@@ -0,0 +1,432 @@
+			return err
+		}
+
+		if err := change(config, changes); err != nil {
+			return err
+		}
+
+		changeBytes, err := proto.Marshal(changes)
+		if err != nil {
+			return err
+		}
+
+		// ...and update the stored changes
+		changeset.Changes = changeBytes
+		if _, err := m.kv.CheckAndSet(csKey, csVersion, changeset); err != nil {
+			if err == kv.ErrVersionMismatch {
+				// Someone else updated the changes first - try again
+				continue
+			}
+
+			return err
+		}
+
+		return nil
+	}
+}
+
+func (m manager) GetPendingChanges() (int, proto.Message, proto.Message, error) {
+	// Get the current configuration, but don't bother trying to create it if it doesn't exist
+	configVal, err := m.kv.Get(m.key)
+	if err != nil {
+		return 0, nil, nil, err
+	}
+
+	// Retrieve the config data so that we can transform it appropriately
+	config := proto.Clone(m.configType)
+	if err := configVal.Unmarshal(config); err != nil {
+		return 0, nil, nil, err
+	}
+
+	// Get the change set for the current configuration, again not bothering to
+	// create if it doesn't exist
+	csKey := fmtChangeSetKey(m.key, configVal.Version())
+	csVal, err := m.kv.Get(csKey)
+	if err != nil {
+		if err == kv.ErrNotFound {
+			// It's ok, just means no pending changes
+			return configVal.Version(), config, nil, nil
+		}
+
+		return 0, nil, nil, err
+	}
+
+	var changeset changesetpb.ChangeSet
+	if err := csVal.Unmarshal(&changeset); err != nil {
+		return 0, nil, nil, err
+	}
+
+	if changeset.State == changesetpb.ChangeSetState_COMMITTED {
+		// There are no pending changes
+		return configVal.Version(), config, nil, nil
+	}
+
+	// Retrieve the changes
+	changes := proto.Clone(m.changesType)
+	if err := proto.Unmarshal(changeset.Changes, changes); err != nil {
+		return 0, nil, nil, err
+	}
+
+	return configVal.Version(), config, changes, nil
+}
+
+func (m manager) Commit(version int, apply ApplyFn) error {
+	// Get the current configuration, but don't bother trying to create it if it doesn't exist
+	configVal, err := m.kv.Get(m.key)
+	if err != nil {
+		return err
+	}
+
+	// Confirm the version does exist...
+	if configVal.Version() < version {
+		return ErrUnknownVersion
+	}
+
+	// ...and that it hasn't already been committed
+	if configVal.Version() > version {
+		return ErrAlreadyCommitted
+	}
+
+	// Retrieve the config data so that we can transform it appropriately
+	config := proto.Clone(m.configType)
+	if err := configVal.Unmarshal(config); err != nil {
+		return err
+	}
+
+	// Get the change set for the current configuration, again not bothering to create
+	// if it doesn't exist
+	csKey := fmtChangeSetKey(m.key, configVal.Version())
+	csVal, err := m.kv.Get(csKey)
+	if err != nil {
+		return err
+	}
+
+	var changeset changesetpb.ChangeSet
+	if err := csVal.Unmarshal(&changeset); err != nil {
+		return err
+	}
+
+	// If the change list is already committed, bail out
+	if changeset.State == changesetpb.ChangeSetState_COMMITTED {
```

Comment:
Say a `Commit` call gets through here and sets the changeset state to committing, but before it bumps up the config version and sets the changeset state to commited, another `Commit` call comes in, sees the changeset state is `ChangeSetState_COMMITTING`, and bumps up the changeset version again. Now the first `Commit` call bumps up the config version, but will fail to set the change set version since it's been updated by the second `Commit` call, and the second `Commit` call will fail when committing the new config since it's been updated by the first `Commit` call.

Do we want to protect ourselves against this case by checking the state against `changesetpb.ChangeSetState_COMMITTING`?